beater/java_attacher/discovery_rules.go

```diff
@@ -0,0 +1,81 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package javaattacher
+
+import (
+	"fmt"
+	"regexp"
+)
+
+type discoveryRule interface {
+	include() bool
+	match(jvm *JvmDetails) bool
+	asString() string
```

Comment:
```suggestion
```

There's a standard interface [`fmt.Stringer`](https://pkg.go.dev/fmt#Stringer). By renaming the `toString` methods on the structs to `String() string`, you'll implement that interface. String formatting functions know about that interface, so there's no need to call the method explicitly. Instead, you can just pass the value in and `%s` will call the `String()` method